src/Workspaces/Core/Portable/FindSymbols/SymbolFinder_Hierarchy.cs

```diff
@@ -184,405 +213,188 @@ public static partial class SymbolFinder
+        /// <param name="solution">The solution to search in.</param>
+        /// <param name="projects">The projects to search. Can be null to search the entire solution.</param>
+#pragma warning disable RS0026 // Do not add multiple public overloads with optional parameters
+        public static Task<IEnumerable<INamedTypeSymbol>> FindImplementationsAsync(
+            INamedTypeSymbol type, Solution solution, IImmutableSet<Project> projects = null, CancellationToken cancellationToken = default)
+#pragma warning restore RS0026 // Do not add multiple public overloads with optional parameters
         {
-            if (symbolToMatch == null)
-            {
-                return false;
-            }
-
-            if (OriginalSymbolsMatchCore(searchSymbol, symbolToMatch, solution, searchSymbolCompilation, symbolToMatchCompilation, cancellationToken))
-            {
-                return true;
-            }
-
-            if (searchSymbol.Kind == SymbolKind.Namespace && symbolToMatch.Kind == SymbolKind.Namespace)
-            {
-                // if one of them is a merged namespace symbol and other one is its constituent namespace symbol, they are equivalent.
-                var namespace1 = (INamespaceSymbol)searchSymbol;
-                var namespace2 = (INamespaceSymbol)symbolToMatch;
-                var namespace1Count = namespace1.ConstituentNamespaces.Length;
-                var namespace2Count = namespace2.ConstituentNamespaces.Length;
-                if (namespace1Count != namespace2Count)
-                {
-                    if ((namespace1Count > 1 &&
-                         namespace1.ConstituentNamespaces.Any(n => NamespaceSymbolsMatch(n, namespace2, solution, cancellationToken))) ||
-                        (namespace2Count > 1 &&
-                         namespace2.ConstituentNamespaces.Any(n2 => NamespaceSymbolsMatch(namespace1, n2, solution, cancellationToken))))
-                    {
-                        return true;
-                    }
-                }
-            }
-
-            return false;
+            return FindImplementationsAsync(type, solution, transitive: true, projects, cancellationToken);
         }
 
-        private static bool OriginalSymbolsMatchCore(
-            ISymbol searchSymbol,
-            ISymbol symbolToMatch,
-            Solution solution,
-            Compilation searchSymbolCompilation,
-            Compilation symbolToMatchCompilation,
-            CancellationToken cancellationToken)
+        /// <summary>
+        /// Finds the accessible <see langword="class"/> or <see langword="struct"/> types that implement the given
+        /// interface.
+        /// </summary>
+        /// <param name="type">The symbol to find derived types of.</param>
+        /// <param name="solution">The solution to search in.</param>
+        /// <param name="transitive">If the search should stop at immediately derived interfaces, or should continue past that.</param>
```

Comment:
It's unclear to me exactly which way the "transitive" flag impacts this. Imagine this:

```
interface IA
{
}

interface IB : IA
{
}

class C : IB
{
}

class D : C
{
}
```

D implements IA only because D inherits C which implements IB which inherits/implements/whatever-we-call-that IA. If I pass transitive "false", which link is being cut here?

Reply:
so, if we're starting with D, we have the following things we can ask it:

1. what is your base type.  i.e. ITypeSymbol.BaseType
2. what are you immediately implemented interfaces.  i.e. ITypeSymbol.Interfaces

if transitive=false we only look at those.

If transitive it 'true':

1. we will look at teh entire .BaseType chain
2. we will look at .AllInterfaces instead of .Interfaces

Reply:
So, in teh above, if you pass transitive=false, and look at 'D', you only get `BaseType=C` and `Interfaces=[<empty>]`